igor-web/src/main/groovy/com/netflix/spinnaker/igor/wercker/WerckerService.groovy

```diff
@@ -0,0 +1,349 @@
+        }
+    }
+
+    String getPipelineId(String appAndPipelineName) {
+        String[] split = appAndPipelineName.split(SPLITOR)
+        String owner = split[0]
+        String appName = split[1]
+        String pipelineName = split[2]
+        String pipelineId = cache.getPipelineID(groupKey, appAndPipelineName)
+        if (pipelineId == null) {
+            try {
+                List<Pipeline> pipelines = werckerClient.getPipelinesForApplication(authHeaderValue, owner, appName)
+                Pipeline matchingPipeline = pipelines.find {pipeline -> pipelineName == pipeline.name}
+                if (matchingPipeline) {
+                    pipelineId = matchingPipeline.id;
+                    cache.setPipelineID(groupKey, appAndPipelineName, pipelineId)
+                }
+            } catch(retrofit.RetrofitError err) {
+                log.info "Error getting pipelines for ${owner} ${appName} ${err} ${err.getClass()}"
+            }
+        }
+        return pipelineId;
+    }
+
+    String getPipelineName(String pipelineId) {
+        String name = cache.getPipelineName(groupKey, pipelineId)
+        if (name == null) {
+            try {
+                Pipeline pipeline = werckerClient.getPipeline(authHeaderValue, pipelineId)
+                if (pipeline && pipeline.application && pipeline.application.owner) {
+                    name = pipeline.application.owner.name + SPLITOR +
+                            pipeline.application.name  + SPLITOR +
+                            pipeline.name
+                    cache.setPipelineID(groupKey, name, pipelineId)
+                }
+            } catch(retrofit.RetrofitError err) {
+                log.info "Error getting pipelines for ${owner} ${appName} ${err} ${err.getClass()}"
+            }
+        }
+        return name;
+    }
+
+    Map<String, List<Run>> getRunsSince(Set<String> pipelines, long since) {
+        long start = System.currentTimeMillis();
+        List<String> pipelineIds = pipelines.collect { getPipelineId(it) };
+        Map<String, String> pipelineKeys = pipelines.collectEntries { [(getPipelineId(it)) : (it)] };
+        List<String> pids = pipelineKeys.keySet().asList();
+        Map<String, List<Run>> pipelineRuns = [:];
+        List<Run> allRuns = werckerClient.getRunsSince(authHeaderValue, branch, pids, limit, since);
+        log.debug "getRunsSince for pipelines:${pipelines} : ${allRuns.size()} runs in ${System.currentTimeMillis() - start}ms!!"
+        allRuns.forEach({ run ->
+            String pipelineKey = pipelineKey(run, pipelineKeys);
+            run.startedAt = run.startedAt?:run.createdAt;
+            List<Run> runs = pipelineRuns.get(pipelineKey);
+            if (runs) {
+                runs.add(run);
+            } else {
+                runs = [run];
+                pipelineRuns.put(pipelineKey, runs);
+            }
+        });
+        return pipelineRuns;
+    }
+
+    Map<String, List<Run>> getRunsSince(long since) {
```

Comment:
This is mostly a duplicate of the `getRunsSince` that accepts `pipelines`.  Can the common logic be factored out?

Reply:
done - moved it to groupRunsByPipeline